arrow/obsidian.c

```diff
@@ -158,18 +170,47 @@ evolve(MTState *random_state,
     // First update all propensities that were affected by the previous event
     for (up = 0; up < update_length; up++) {
 
-      // Find which reaction to update and initialize the propensity for that reaction
-      // with the rate for that reaction
+      // Find which reaction to update
       reaction = update[up];
-      propensities[reaction] = rates[reaction];
-
-      // Go through each reactant and calculate its contribution to the propensity
-      // based on the counts of the corresponding substrate, which is then multiplied
-      // by the reaction's original rate and the contributions from other reactants
-      for (reactant = 0; reactant < reactants_lengths[reaction]; reactant++) {
-        index = reactants_indexes[reaction] + reactant;
-        count = outcome[reactants[index]];
-        propensities[reaction] *= choose(count, reactions[index]);
+
+      // Compute propensity depending on the reaction's form
+      switch(forms[reaction]) {
+
+        case 0: // Standard Gillespie
+          // Initialize the propensity for reaction with its rate constant
+          propensities[reaction] = rates_flat[reaction];
+
+          // Go through each reactant and calculate its contribution to the
+          // propensity based on the counts of the corresponding substrate,
+          // which is then multiplied by the reaction's original rate and the
+          // contributions from other reactants
+          for (reactant = 0; reactant < reactants_lengths[reaction]; reactant++) {
+            index = reactants_indexes[reaction] + reactant;
+            count = outcome[reactants[index]];
+            propensities[reaction] *= choose(count, reactions[index]);
+          }
+
+          break;
+
+        case 1: // Michaelis-Menten
```

Comment:
Right, the `form` switch. We may want to talk about what it looks like to 

* Add more forms.
* Use user provided forms.
* Move the essence of each form into its own function?

There will get to be a point where this switch gets unwieldy, but it serves for now.